lib/viewmodels/challenge_view_model.dart

```diff
@@ -1,36 +1,90 @@
-        distance: 3200,
-        timeLeft: 27,
-        reward: 400,
-      ),
-      ChallengeCardData.group(
-        title: "I'm moving out",
-        distance: 4000,
-        reward: 350,
-      ),
-      ChallengeCardData.soloFinished(
-        title: "What a view!",
-        timeLeft: 27,
-        reward: 200,
-      ),
-    ];
+    final geoLocationService = ref.watch(geoLocationServiceProvider);
+    final challengeRepository = ref.watch(challengeRepositoryServiceProvider);
+
+    final currentPosition = await geoLocationService.getCurrentPosition();
+    final currentUser = ref.watch(uidProvider);
+
+    final firestoreChallenges = await challengeRepository.getChallenges(
+      currentUser!,
+      currentPosition,
+    );
+
+    final postRepository = ref.watch(postRepositoryProvider);
+    final now = DateTime.now();
+    Iterable<Future<ChallengeCardData>> futureUiChallenges =
+        firestoreChallenges.map((challenge) async {
+      final post = await postRepository.getPost(challenge.postId);
+      final timeLeft = challenge.data.expiresOn.toDate().difference(now);
+
+      if (!challenge.data.isCompleted) {
+        final double distanceKm = GeoFirePoint(currentPosition)
+            .distanceBetweenInKm(geopoint: post.location.geoPoint);
+        final int distanceM = (distanceKm * 1000).toInt();
+
+        return ChallengeCardData.solo(
+          title: post.data.title,
+          distance: distanceM,
+          timeLeft: timeLeft.inHours,
+          reward: ChallengeRepositoryService.soloChallengeReward,
+        );
+      } else {
+        return ChallengeCardData.soloFinished(
+          title: post.data.title,
+          timeLeft: timeLeft.inHours,
+          reward: ChallengeRepositoryService.soloChallengeReward,
+        );
+      }
+    });
+
+    final uiChallenges = await Future.wait(futureUiChallenges);
+    uiChallenges.sort((a, b) {
+      if (b.isFinished) {
+        return -1;
+      } else {
+        return 1;
+      }
+    }); // sort the list so that finished challenges appear last
+
+    return uiChallenges;
+  }
+
+  /// Refresh the list of challenges
+  Future<void> refresh() async {
+    state = const AsyncValue.loading();
+    state = await AsyncValue.guard(() => build());
+  }
+
+  /// Complete a challenge associated with post of id [pid]. Returns true if the
+  /// challenge was indeed completed and points were awarded.
+  /// Returns false if the challenge could not be
+  /// completed. This could be because the post given was not a challenge, was
+  /// expired, or was already completed.
+  Future<bool> completeChallenge(PostIdFirestore pid) async {
+    final currentUser = ref.read(uidProvider);
+    final challengeRepository = ref.read(challengeRepositoryServiceProvider);
+
+    final challengeCompleted =
+        await challengeRepository.completeChallenge(currentUser!, pid);
+    await refresh();
```

Comment:
I did that in 6b9d5ed2069fe121907fb1327195b2c1088254aa. I had to add a delay in the tests to wait for the view-model to refresh. I also made it only refresh when a challenge actually completed, since this is called when clicking on every post.

Reply:
Ok, so this is purely for testing. I'm not sure it makes the app faster, I'm guessing it just makes the test await correctly, without necessary supplementary delay.
This allows to this more easily of the code, and is therefore less error-prone. Let's keep it this way! :)

Reply:
e6cfcac3565d907b31ee7d11f6a501b9638acbbc